pkg/client/v1/nimbleos/access_control_record.go

```diff
@@ -0,0 +1,133 @@
+	// InitiatorGroupName - Name of the initiator group.
+	InitiatorGroupName *string `json:"initiator_group_name,omitempty"`
+	// Lun - If this access control record applies to a regular volume, this attribute is the volume's LUN (Logical Unit Number). If the access protocol is iSCSI, the LUN will be 0. However, if the access protocol is Fibre Channel, the LUN will be in the range from 0 to 2047. If this record applies to a Virtual Volume, this attribute is the volume's secondary LUN in the range from 0 to 399999, for both iSCSI and Fibre Channel. If the record applies to a OpenstackV2 volume, the LUN will be in the range from 0 to 2047, for both iSCSI and Fibre Channel. If this record applies to a protocol endpoint or only a snapshot, this attribute is not meaningful and is set to null.
+	Lun *int64 `json:"lun,omitempty"`
+	// VolId - Identifier for the volume this access control record applies to.
+	VolId *string `json:"vol_id,omitempty"`
+	// VolName - Name of the volume this access control record applies to.
+	VolName *string `json:"vol_name,omitempty"`
+	// VolAgentType - External management agent type.
+	VolAgentType *NsAgentType `json:"vol_agent_type,omitempty"`
+	// PeId - Identifier for the protocol endpoint this access control record applies to.
+	PeId *string `json:"pe_id,omitempty"`
+	// PeName - Name of the protocol endpoint this access control record applies to.
+	PeName *string `json:"pe_name,omitempty"`
+	// PeLun - LUN (Logical Unit Number) to associate with this protocol endpoint. Valid LUNs are in the 0-2047 range.
+	PeLun *int64 `json:"pe_lun,omitempty"`
+	// SnapId - Identifier for the snapshot this access control record applies to.
+	SnapId *string `json:"snap_id,omitempty"`
+	// SnapName - Name of the snapshot this access control record applies to.
+	SnapName *string `json:"snap_name,omitempty"`
+	// PeIds - List of candidate protocol endpoints that may be used to access the Virtual Volume. One of them will be selected for the access control record. This field is required only when creating an access control record for a Virtual Volume.
+	PeIds []*string `json:"pe_ids,omitempty"`
+	// Snapluns - Information about the snapshot LUNs associated with this access control record. This field is meaningful when the online snapshot can be accessed as a LUN in the group.
+	Snapluns []*NsSnapLunInfo `json:"snapluns,omitempty"`
+	// CreationTime - Time when this access control record was created.
+	CreationTime *int64 `json:"creation_time,omitempty"`
+	// LastModified - Time when this access control record was last modified.
+	LastModified *int64 `json:"last_modified,omitempty"`
+	// AccessProtocol - Access protocol of the volume.
+	AccessProtocol *NsAccessProtocol `json:"access_protocol,omitempty"`
+}
+
+// EncodeAccessControlRecord - Transform AccessControlRecord to accessControlRecord type
+func EncodeAccessControlRecord(request interface{}) (*accessControlRecord, error) {
+	reqAccessControlRecord := request.(*AccessControlRecord)
```

Comment:
No, there are 45 different types of objects, and writing a type assertion for each of them doesn't add any value. Although, I do need to add err check in the Encoder and Decoder which I have missed.

Reply:
In EncodeAccessControlRecord() it can be done but again in DecodeAccessControlRecord() it cant be done. That will break the consistency . Since, during EncodeAccessControlRecord this will accepts concrete data type and during DecodeAccessControlRecord() an generic data type.  Moreover, we still this type resp== nil check, it will just move to EncodeAccessControlRecord().

Reply:
The Encode and Decode functions look symmetric to me.  A function named "DecodeAccessControlRecord" would seem to be specific to the accessControlRecord type.  Why does it need to be generic?

Reply:
Because we dont know the POST return type until we land in the specific object set . The implementation of POST in group_mgmt_client.go is generic . please check .